sno/working_copy/mysql.py

```diff
@@ -0,0 +1,327 @@
+                {"table_name": dataset.table_name},
+            )
+        )
+
+    def _drop_triggers(self, sess, dataset):
+        sess.execute(f"DROP TRIGGER {self._quoted_trigger_name(dataset, 'ins')}")
+        sess.execute(f"DROP TRIGGER {self._quoted_trigger_name(dataset, 'upd')}")
+        sess.execute(f"DROP TRIGGER {self._quoted_trigger_name(dataset, 'del')}")
+
+    @contextlib.contextmanager
+    def _suspend_triggers(self, sess, dataset):
+        self._drop_triggers(sess, dataset)
+        yield
+        self._create_triggers(sess, dataset)
+
+    def meta_items(self, dataset):
+        with self.session() as sess:
+            table_info_sql = """
+                SELECT
+                    C.column_name, C.ordinal_position, C.data_type, C.srs_id,
+                    C.character_maximum_length, C.numeric_precision, C.numeric_scale,
+                    KCU.ordinal_position AS pk_ordinal_position
+                FROM information_schema.columns C
+                LEFT OUTER JOIN information_schema.key_column_usage KCU
+                ON (KCU.table_schema = C.table_schema)
+                AND (KCU.table_name = C.table_name)
+                AND (KCU.column_name = C.column_name)
+                WHERE C.table_schema=:table_schema AND C.table_name=:table_name
+                ORDER BY C.ordinal_position;
+            """
+            r = sess.execute(
+                table_info_sql,
+                {"table_schema": self.db_schema, "table_name": dataset.table_name},
+            )
+            mysql_table_info = list(r)
+
+            spatial_ref_sys_sql = """
+                SELECT SRS.* FROM information_schema.st_spatial_reference_systems SRS
+                LEFT OUTER JOIN information_schema.st_geometry_columns GC ON (GC.srs_id = SRS.srs_id)
+                WHERE GC.table_schema=:table_schema AND GC.table_name=:table_name;
+            """
+            r = sess.execute(
+                spatial_ref_sys_sql,
+                {"table_schema": self.db_schema, "table_name": dataset.table_name},
+            )
+            mysql_spatial_ref_sys = list(r)
+
+            id_salt = f"{self.db_schema} {dataset.table_name} {self.get_db_tree()}"
+            schema = mysql_adapter.sqlserver_to_v2_schema(
+                mysql_table_info, mysql_spatial_ref_sys, id_salt
+            )
+            yield "schema.json", schema.to_column_dicts()
+
+    @classmethod
+    def try_align_schema_col(cls, old_col_dict, new_col_dict):
+        old_type = old_col_dict["dataType"]
+        new_type = new_col_dict["dataType"]
+
+        # Some types have to be approximated as other types in MySQL
+        if mysql_adapter.APPROXIMATED_TYPES.get(old_type) == new_type:
+            new_col_dict["dataType"] = new_type = old_type
+            for key in mysql_adapter.APPROXIMATED_TYPES_EXTRA_TYPE_INFO:
+                new_col_dict[key] = old_col_dict.get(key)
+
+        # Geometry types don't have to be approximated, except for the Z/M specifiers.
```

Comment:
Z and M values are allowed, since MySQL basically stores WKB internally - but AFAICT, there's no way to specify a column as containing only 2, 3, or 4 dimensional geometries. Since Kart's schema does contain this metadata, this schema information is "approximated" - ie a column restricted to n-dimensional geometries where n is ONE OF (2, 3, or 4) in Kart is approximated as a column for n-dimensional geometries, where n is ANY/ALL of (2,3,4) in MySQL.

Reply:
Added a comment.

Reply:
Z and M values are *not* allowed. Should've checked that. New plan - don't support datasets with Z or M values for now. Discussion of other longer-term solutions can be found in the slack channel.